src/machine/asm/mod.rs

```diff
@@ -303,14 +303,17 @@ impl SupportMachine for Box<AsmCoreMachine> {
         Some(self.max_cycles)
     }
 
-    fn reset(&mut self) {
+    fn reset(&mut self, max_cycles: Option<u64>) {
         self.registers = [0; RISCV_GENERAL_REGISTER_NUMBER];
         self.pc = 0;
         self.flags = [0; RISCV_PAGES];
         for i in 0..TRACE_SIZE {
             self.traces[i] = Trace::default();
         }
         self.frames = [0; MEMORY_FRAMES];
+        self.cycles = 0;
+        debug_assert!(max_cycles.is_some());
+        self.max_cycles = max_cycles.unwrap_or(0);
```

Comment:
https://github.com/nervosnetwork/ckb-vm/blob/develop/definitions/src/asm.rs#L61

Inconsistent with the default behavior

By the way, why asm machine defaults to u64 max and rust machine defaults to 0

Reply:
1) The purpose here is to ensure that max_cycles is not None, and unwrap_or(0) can be safely ignored.

2) See: https://github.com/nervosnetwork/ckb-vm/blob/develop/definitions/src/asm.rs#L58-L60

Reply:
This is weird to me, I think we should stick to either one:

* ensure users always provide a max cycle in the reset call here
* or fix the machine, so it can have max cycle or not

This `unwrap_or` call here, does not make too much sense to me.

Reply:
Since AsmMachine always needs to set cycles, I think the option 1) is better.

Reply:
Maybe this solution is better:
```
if let Some(cycles) = max_cycles {
  self.max_cycles = cycles;
}
```

Reply:
Accept.

Reply:
I do have a concern regarding this solution: at a lower level, AsmMachine does not accept an option, you should always have a max cycles set, one typical solution, is to use `u64::max_value()` in case one does not require a max cycle. This brings a problem: when we are using None in reset function, what behavior are we achieving? We now have 2 different APIs:

* When you are initializing a VM, you fill in `u64::max_value()` when you do not have a cycle to set
* When you are resetting, you can use `None` or `u64::max_value()`

In this sense, I'm not sure using Option here, is a good solution.

Reply:
I think most users will use the new method to create an asm machine, reset with None means keep the original max cycles

Reply:
Option is necessary because RustMachine requires it. 

https://github.com/nervosnetwork/ckb-vm/blob/develop/src/machine/mod.rs#L274

Reply:
The root problem here, is that the original max cycles are kept, so we are just introducing a new rule to remember here. The Option here, does not actually provide what a typical Option value provides. In this sense, I would more be willing to remove Option here and always require a value. If a user does not have a proper value to set, they can just use `u64::max_value()`

Reply:
For RustMachine's reset, None means that the total cycle size is not limited.

For AsmMachine, None should perform a similar function. So if it is None, is it better to use u64::max_value()? 

```rs
max_cycles.unwrap_or(u64::max_value)
```

Reply:
I think we should just have consistent behavior.

If we want to support Option, we should support it in both places, when you are initializing a VM, and when you are resetting. If we cannot support it, we should not provide an Option here.

Reply:
Part of me think it might be better just to ensure you always have a max cycles set, and remove that Option in RustMachine.

If one does not care for cycles, they can always choose not to provide a cost function, or a function that always return 0. That to me will be a better solution.

Reply:
> Part of me think it might be better just to ensure you always have a max cycles set, and remove that Option in RustMachine.

It is a perfect solution.